### ISSUE TITLE
HOST max_check_attempts set to 1 goes SOFT and then HARD, extra check is performed

### ISSUE THREAD

**Issue by box293** [NagiosEnterprises/nagioscore#23]
Originally reported here http://support.nagios.com/forum/viewtopic.php?f=7&t=31262&p=126275#p125754

Nagios Core 4.0.8 on CentOS 6.5

With HOST object definitions, when the max_check_attempts is set to 1, when a host goes down, it enters a SOFT state first and then a HARD state, resulting in two checks performed instead of one.

SERVICE object definitions behave correctly.

Here is an extract from a nagios.debug demonstrating this behaviour:

```
HOST Up then Down then Up

[1423605101.333215] [016.0] [pid=14074] Scheduling a forced, active check of host 'centos01' @ Wed Feb 11 08:51:38 2015
[1423605101.333277] [008.0] [pid=14074] ** Timed Event ** Type: EVENT_HOST_CHECK, Run Time: Wed Feb 11 08:51:38 2015
[1423605101.333279] [008.0] [pid=14074] ** Host Check Event ==> Host: 'centos01', Options: 1, Latency: 0.000050 sec
[1423605101.338902] [016.1] [pid=14074] HOST: centos01, ATTEMPT=1/1, CHECK TYPE=ACTIVE, STATE TYPE=HARD, OLD STATE=0, NEW STATE=0
[1423605101.338911] [016.1] [pid=14074] Host was UP.
[1423605101.338917] [016.1] [pid=14074] Host is still UP.
[1423605101.338924] [016.1] [pid=14074] Pre-handle_host_state() Host: centos01, Attempt=1/1, Type=HARD, Final State=0 (UP)
[1423605101.339083] [016.1] [pid=14074] Post-handle_host_state() Host: centos01, Attempt=1/1, Type=HARD, Final State=0 (UP)
[1423605101.339297] [016.0] [pid=14074] Scheduling a non-forced, active check of host 'centos01' @ Wed Feb 11 08:56:41 2015

[1423605136.947174] [016.0] [pid=14074] Scheduling a forced, active check of host 'centos01' @ Wed Feb 11 08:52:10 2015
[1423605136.947324] [008.0] [pid=14074] ** Timed Event ** Type: EVENT_HOST_CHECK, Run Time: Wed Feb 11 08:52:10 2015
[1423605136.947329] [008.0] [pid=14074] ** Host Check Event ==> Host: 'centos01', Options: 1, Latency: 0.000134 sec
[1423605146.953117] [016.1] [pid=14074] HOST: centos01, ATTEMPT=1/1, CHECK TYPE=ACTIVE, STATE TYPE=HARD, OLD STATE=0, NEW STATE=1
[1423605146.953121] [016.1] [pid=14074] Host was UP.
[1423605146.953124] [016.1] [pid=14074] Host is now UP.
[1423605146.953134] [016.1] [pid=14074] Pre-handle_host_state() Host: centos01, Attempt=1/1, Type=SOFT, Final State=1 (DOWN)
[1423605146.953417] [016.1] [pid=14074] Post-handle_host_state() Host: centos01, Attempt=1/1, Type=SOFT, Final State=1 (DOWN)
[1423605146.953462] [016.0] [pid=14074] Scheduling a non-forced, active check of host 'centos01' @ Wed Feb 11 08:53:26 2015

[1423605206.929373] [008.0] [pid=14074] ** Timed Event ** Type: EVENT_HOST_CHECK, Run Time: Wed Feb 11 08:53:26 2015
[1423605206.929383] [008.0] [pid=14074] ** Host Check Event ==> Host: 'centos01', Options: 0, Latency: 0.000000 sec
[1423605209.933987] [016.1] [pid=14074] HOST: centos01, ATTEMPT=1/1, CHECK TYPE=ACTIVE, STATE TYPE=SOFT, OLD STATE=1, NEW STATE=1
[1423605209.934000] [016.1] [pid=14074] Host was DOWN.
[1423605209.934010] [016.1] [pid=14074] Host is still DOWN.
[1423605209.934031] [016.1] [pid=14074] Pre-handle_host_state() Host: centos01, Attempt=1/1, Type=HARD, Final State=1 (DOWN)
[1423605209.938604] [016.1] [pid=14074] Post-handle_host_state() Host: centos01, Attempt=1/1, Type=HARD, Final State=1 (DOWN)
[1423605209.938683] [016.0] [pid=14074] Scheduling a non-forced, active check of host 'centos01' @ Wed Feb 11 08:58:29 2015

[1423605294.255760] [016.0] [pid=14074] Scheduling a forced, active check of host 'centos01' @ Wed Feb 11 08:54:51 2015
[1423605294.255831] [008.0] [pid=14074] ** Timed Event ** Type: EVENT_HOST_CHECK, Run Time: Wed Feb 11 08:54:51 2015
[1423605294.255832] [008.0] [pid=14074] ** Host Check Event ==> Host: 'centos01', Options: 1, Latency: 0.000055 sec
[1423605294.259083] [016.1] [pid=14074] HOST: centos01, ATTEMPT=1/1, CHECK TYPE=ACTIVE, STATE TYPE=HARD, OLD STATE=1, NEW STATE=0
[1423605294.259085] [016.1] [pid=14074] Host was DOWN.
[1423605294.259086] [016.1] [pid=14074] Host experienced a HARD recovery (it's now UP).
[1423605294.259099] [016.1] [pid=14074] Pre-handle_host_state() Host: centos01, Attempt=1/1, Type=HARD, Final State=0 (UP)
[1423605294.259394] [016.1] [pid=14074] Post-handle_host_state() Host: centos01, Attempt=1/1, Type=HARD, Final State=0 (UP)


Service Up then Down then Up

[1423605375.013349] [016.0] [pid=15447] Scheduling a forced, active check of service 'SSH Server' on host 'centos01' @ Wed Feb 11 08:56:12 2015
[1423605375.042681] [016.1] [pid=15447] Service is OK.
[1423605375.042683] [016.1] [pid=15447] Service did not change state.
[1423605375.042711] [016.0] [pid=15447] Scheduling a non-forced, active check of service 'SSH Server' on host 'centos01' @ Wed Feb 11 09:01:15 2015

[1423605389.043332] [016.0] [pid=15447] Scheduling a forced, active check of service 'SSH Server' on host 'centos01' @ Wed Feb 11 08:56:23 2015
[1423605389.083359] [016.1] [pid=15447] Service is in a non-OK state!
[1423605389.083361] [016.1] [pid=15447] Host is currently UP, so we'll recheck its state to make sure...
[1423605389.083366] [001.0] [pid=15447] schedule_host_check()
[1423605389.083383] [016.0] [pid=15447] Scheduling a non-forced, active check of host 'centos01' @ Wed Feb 11 08:56:29 2015
[1423605389.083443] [016.1] [pid=15447] Current/Max Attempt(s): 1/1
[1423605389.083447] [016.1] [pid=15447] Service has reached max number of rechecks, so we'll handle the error...
[1423605389.083937] [016.0] [pid=15447] Scheduling a non-forced, active check of service 'SSH Server' on host 'centos01' @ Wed Feb 11 09:01:29 2015

[1423605431.622650] [016.0] [pid=15447] Scheduling a forced, active check of service 'SSH Server' on host 'centos01' @ Wed Feb 11 08:57:03 2015
[1423605431.665858] [016.1] [pid=15447] Service is OK.
[1423605431.665860] [016.1] [pid=15447] Service experienced a HARD RECOVERY.
[1423605431.666198] [016.0] [pid=15447] Scheduling a non-forced, active check of service 'SSH Server' on host 'centos01' @ Wed Feb 11 09:02:11 2015
```

**Comment by tmcnag** [comment]
Implemented and tested the fix in 4.1, but observed that the buggy behavior was present in both 4.0.8 and 4.1 so it made no sense fixing it for 4.0.8.  Didn't notice any side effects, but it was on a toy system so there might be some interesting ramifications e.g. notifications and event handlers.

**Comment by jfrickson** [comment]
In discussion with tmcnag and scottwilkerson, we decided that this change has potential wide-ranging effects, so it should go in (if at all) with a major release, such as 5.0. The issue will remain open, but the pull request has been closed.

**Comment by tmcnag** [comment]
To clarify, the changes would mostly be related to escalation and event handler logic.  Implementing my change would potentially introduce an off-by-one situation with certain installs that use the aforementioned features.  Someone expecting the HARD state to come after the SOFT state might have implemented an event handler that deals with the SOFT state in a certain way, and this change would prevent that from ever happening.

In keeping this issue open, if anyone is interested I can always make this a configuration option.  This would increase config file bloat a bit, but if there is enough interest it might be worth discussing.

**Comment by Coservit** [comment]
Hi @tmcnag ,

We develop a monitoring solution and we have about 2 thousand nagios servers in production with a centralized configuration UI . We are updating from nagios 3.5 to nagios 4.2.1. The change that @box293 is talking about isn't written in any release note (at least I don't see it...). So, the change have silently modified the behavior of all updated nagios servers already in production.

Can you please explain in more details why we have to wait for a major version (5.0?)?
Those updating from nagios 3.5 would have a bad surprise when they discover this. Our configuration UI is now incoherent with the updated behavior (it is now impossible to disable retry attempt even if we set the max_check_attemps to 1. In that case there should be only one attempt without retry).

I think it should be modified soon in a minor version. Or at least, please "make this a configuration option" to keep the nagios 3.5 behavior.

Thank you.

**Comment by Coservit** [comment]
@tmcnag if you need any assistance from us for this issue, we are at your disposal.

**Comment by tmcnag** [comment]
@Coservit For clarification, the changes I made were to my own repo and were never merged into Core proper.  I never even made the pull request.  If you are experiencing the described behavior, you must be running my fork of Core.

**Comment by scottwilkerson** [comment]
This was never addressed, just closed by jfrickson

I believe this [forum post](https://support.nagios.com/forum/viewtopic.php?t=47685#p245937) is directly related and @tmcnag's fix would solve the issue
https://github.com/tmcnag/nagioscore/commit/2f263165afa691151ee1f26c878e35ddf2db0972

**Comment by hedenface** [comment]
This has been fixed in 4.4.0. I forget the commit exactly, but you can see it in base/checks.c in the function `host_state_or_hard_state_type_change`.
